src/java/org/apache/cassandra/io/tries/IncrementalTrieWriterPageAware.java

```diff
@@ -0,0 +1,715 @@
+
+        int actualSize = recalcTotalSizeRecursive(root, dest.position());
+        int bytesLeft = dest.bytesLeftInPage();
+        if (actualSize > bytesLeft)
+        {
+            if (actualSize <= maxBytesPerPage)
+            {
+                dest.padToPageBoundary();
+                bytesLeft = maxBytesPerPage;
+                // position changed, recalculate again
+                actualSize = recalcTotalSizeRecursive(root, dest.position());
+            }
+
+            if (actualSize > bytesLeft)
+            {
+                // Still greater. Lay out children separately.
+                layoutChildren(root);
+
+                // Pad if needed and place.
+                if (root.nodeSize > dest.bytesLeftInPage())
+                {
+                    dest.padToPageBoundary();
+                    // Recalculate again as pointer size may have changed, triggering assertion in writeRecursive.
+                    recalcTotalSizeRecursive(root, dest.position());
+                }
+            }
+        }
+
+
+        root.finalizeWithPosition(writeRecursive(root));
+        return root;
+    }
+
+    @Override
+    void complete(Node<Value> node) throws IOException
+    {
+        assert node.filePos == -1;
+
+        int branchSize = 0;
+        for (Node<Value> child : node.children)
+            branchSize += child.branchSize + child.nodeSize;
+
+        node.branchSize = branchSize;
+
+        int nodeSize = serializer.sizeofNode(node, dest.position());
+        if (nodeSize + branchSize < maxBytesPerPage)
+        {
+            // Good. This node and all children will (most probably) fit page.
+            node.nodeSize = nodeSize;
+            node.hasOutOfPageChildren = false;
+            node.hasOutOfPageInBranch = false;
+
+            for (Node<Value> child : node.children)
+                if (child.filePos != -1)
+                    node.hasOutOfPageChildren = true;
+                else if (child.hasOutOfPageChildren || child.hasOutOfPageInBranch)
+                    node.hasOutOfPageInBranch = true;
+
+            return;
+        }
+
+        // Cannot fit. Lay out children; The current node will be marked with one with out-of-page children.
+        layoutChildren(node);
+    }
+
+    private void layoutChildren(Node<Value> node) throws IOException
+    {
+        assert node.filePos == -1;
+
+        NavigableSet<Node<Value>> children = new TreeSet<>(BRANCH_SIZE_COMPARATOR);
+        for (Node<Value> child : node.children)
+            if (child.filePos == -1)
+                children.add(child);
+
+        int bytesLeft = dest.bytesLeftInPage();
+        Node<Value> cmp = new Node<>(256); // goes after all equal-sized unplaced nodes (whose transition character is 0-255)
+        cmp.nodeSize = 0;
+        while (!children.isEmpty())
+        {
+            cmp.branchSize = bytesLeft;
+            Node<Value> child = children.headSet(cmp, true).pollLast();    // grab biggest that could fit
```

Comment:
I though that rather `floor`? Also this is not equivalent - `pollLast` removes the retrieved element - I thought that it is a part of the algorithm since the loop iterates until `children` is empty?

Reply:
You are right, sorry about this.